### PR TITLE
Account-Level Defaults for Cookie Sync Endpoint

### PR DESCRIPTION
**Pull request by AlexBVolcy**
Addressing this issue: https://github.com/prebid/prebid-server/issues/1641

This PR adds Account-Level Defaults to the cookie-sync endpoint

The defaults are `DefaultLimit`, `MaxLimit`, and `DefaultCoopSync`, and were added to the `Account` struct in `config/accounts.go`

The defaults are utilized in this manner:
- If a cookie sync request doesn't contain a `limit` field, we then set the field to equal the `DefaultLimit`
- If a cookie sync request contains a `limit` field, but it's larger than the `MaxLimit`, then we set the field equal to `MaxLimit`
- If a cookie sync request doesn't contain a `coopSync` field, we set the field equal to `DefaultCoopSync`

The `TestCookieSyncParseRequest` has been updated with more cases, along with tweaks to existing cases, to support the changes.

### CODE REVIEW

File: config/accounts.go [hunk 1]
```diff
@@ -21,6 +21,9 @@ type Account struct {
 	GDPR               AccountGDPR `mapstructure:"gdpr" json:"gdpr"`
 	DebugAllow         bool        `mapstructure:"debug_allow" json:"debug_allow"`
 	DefaultIntegration string      `mapstructure:"default_integration" json:"default_integration"`
+	DefaultLimit       int         `mapstructure:"default_limit" json:"default_limit"`
+	MaxLimit           int         `mapstructure:"max_limit" json:"max_limit"`
+	DefaultCoopSync    bool        `mapstructure:"default_coop_sync" json:"default_coop_sync"`
```
Comment (SyntaxNode) [inline config/accounts.go:26]: I recommend creating an object structure to provide context and further extension points in the future, such as embedding these three within a "UserSync" object.

File: endpoints/cookie_sync.go [hunk 2]
```diff
@@ -63,6 +63,7 @@ func NewCookieSyncEndpoint(
 type cookieSyncEndpoint struct {
 	chooser          usersync.Chooser
 	config           config.UserSync
+	accountConfig    config.Account
```
Comment (SyntaxNode) [inline endpoints/cookie_sync.go:66]: This data structure is for the life of the application, but  account config should be for the life of the request. You'll want to add the account as a string in the `cookieSyncRequest` struct and use that with the `stored_requests.AccountFetcher` to lookup the account config. So it's really the AccountFetcher you want to include at this level.

There is also no automatic dependency injection in this project. You'll need to wire that up manually in router.go#New. Look for the line:

```
r.POST("/cookie_sync", endpoints.NewCookieSyncEndpoint(syncersByBidder, cfg, gdprPerms, r.MetricsEngine, pbsAnalytics, activeBidders).Handle)
```

where you'll have the `accounts` variable for the AccountFetcher in scope.

File: config/accounts.go (hunk 1 above)
Comment (mansinahar) [reply config/accounts.go:26]: I agree with @SyntaxNode. I'd just say that we should probably use `cookie_sync` as the wrapper object for these fields so as to stay consistent with naming in Java.
Ref: https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/settings/model/Account.java#L23

File: endpoints/cookie_sync_test.go [hunk 3]
```diff
@@ -647,18 +657,109 @@ func TestCookieSyncParseRequest(t *testing.T) {
...
+		{
+			description: "Request where account level defaults are applied: - MaxLimit and Default Coop",
+			givenBody: strings.NewReader(`{` +
+				`"bidders":["a", "b"],` +
+				`"gdpr":1,` +
+				`"gdpr_consent":"anyGDPRConsent",` +
+				`"us_privacy":"1NYN",` +
+				`"limit":42,` +
+				`"account":"TestAccount"` +
+				`}`),
```
Comment (SyntaxNode) [inline endpoints/cookie_sync_test.go:669]: I recommend removing gdpr, gdpr_consent, and us_privacy from the request. It's ideal for the test setup to be as simple as possible and these values don't seem to have an impact on the area of code you are testing. Look to a test case like "Cooperative True - Default False" as an example.

File: endpoints/cookie_sync.go [hunk 4]
```diff
@@ -125,6 +132,21 @@ func (c *cookieSyncEndpoint) parseRequest(r *http.Request) (usersync.Request, pr
...
+		}
+		if request.Limit == 0 {
+			request.Limit = accountInfo.Defaults.DefaultLimit
+		} else if request.Limit > accountInfo.Defaults.MaxLimit {
+			request.Limit = accountInfo.Defaults.MaxLimit
+		}
+		if request.CooperativeSync == nil {
+			request.CooperativeSync = &accountInfo.Defaults.DefaultCoopSync
+		}
+	}
```
Comment (SyntaxNode) [inline endpoints/cookie_sync.go:148]: I like the tests you have in `TestNewCookieSyncEndpoint` for end to end functionality. Please also add tests to `TestCookieSyncParseRequest` which can include additional permutations, such as handling a potential error from the account fetcher on line 136.

**Review by VeronikaSolovei9 — COMMENTED** [review]
Overall LGTM, please address SyntaxNode's comments

File: endpoints/cookie_sync.go [hunk 5]
```diff
@@ -47,26 +50,30 @@ func NewCookieSyncEndpoint(
 
 	return &cookieSyncEndpoint{
 		chooser:          usersync.NewChooser(syncersByBidder),
-		config:           config.UserSync,
+		config:           config,
+		configUserSync:   config.UserSync,
```
[the remark this answers is not in the document]
Comment (AlexBVolcy) [reply endpoints/cookie_sync.go:54]: Do you think we should alter the cookie sync endpoint struct to just include a config field now as well?

File: endpoints/cookie_sync.go [hunk 6]
```diff
@@ -125,6 +132,21 @@ func (c *cookieSyncEndpoint) parseRequest(r *http.Request) (usersync.Request, pr
 		}
 	}
 
+	if request.Account != "" {
+		accountInfo, errs := accountService.GetAccount(context.Background(), c.config, c.accountsFetcher, request.Account)
+		if len(errs) > 0 {
+			return usersync.Request{}, privacy.Policies{}, errs[0]
```
[the remark this answers is not in the document]
Comment (AlexBVolcy) [reply endpoints/cookie_sync.go:138]: Yeah I was also confused about this, because account service returns an array of errors, but the parse request function only returns a single error. Veronika, are you suggesting we refactor the GetAccount function?

File: endpoints/cookie_sync.go [hunk 7]
```diff
@@ -257,6 +258,15 @@ func (c *cookieSyncEndpoint) handleError(w http.ResponseWriter, err error, httpS
 	})
 }
 
+func combineErrors(errs []error) error {
+	var errorStringList []string
```
Comment (SyntaxNode) [inline endpoints/cookie_sync.go:262]: The `List` at the end of the variable name looks like Hungarian notation. Please consider removing it.

File: endpoints/cookie_sync.go [hunk 8]
```diff
@@ -125,6 +131,21 @@ func (c *cookieSyncEndpoint) parseRequest(r *http.Request) (usersync.Request, pr
...
+			return usersync.Request{}, privacy.Policies{}, combineErrors(errs)
+		}
+		if request.Limit == 0 {
+			request.Limit = accountInfo.CookieSync.DefaultLimit
+		} else if request.Limit > accountInfo.CookieSync.MaxLimit {
+			request.Limit = accountInfo.CookieSync.MaxLimit
+		}
+		if request.CooperativeSync == nil {
+			request.CooperativeSync = &accountInfo.CookieSync.DefaultCoopSync
+		}
```
Comment (SyntaxNode) [inline endpoints/cookie_sync.go:146]: There is a logic gap here. As a publisher, I expect my max limit to always be honored. However, if you have a 0 limit and a 0 default limit, the limit will remain at 0 which is interpreted as unlimited. I had luck with the following:

```
	if request.Account != "" {
		accountInfo, errs := accountService.GetAccount(context.Background(), c.config, c.accountsFetcher, request.Account)
		if len(errs) > 0 {
			return usersync.Request{}, privacy.Policies{}, combineErrors(errs)
		}

		if request.Limit == 0 {
			request.Limit = accountInfo.CookieSync.DefaultLimit
		}

		if request.Limit == 0 || request.Limit > accountInfo.CookieSync.MaxLimit {
			request.Limit = accountInfo.CookieSync.MaxLimit
		}

		if request.CooperativeSync == nil {
			request.CooperativeSync = &accountInfo.CookieSync.DefaultCoopSync
		}
	}
```

**Review by VeronikaSolovei9 — APPROVED** [review]
LGTM

**Review by guscarreon — COMMENTED** [review]
Looks good to me. Left a couple of nitpicks

**Review by VeronikaSolovei9 — DISMISSED** [review]
Looks good

File: endpoints/cookie_sync.go [hunk 9]
```diff
@@ -125,6 +132,22 @@ func (c *cookieSyncEndpoint) parseRequest(r *http.Request) (usersync.Request, pr
...
+		accountInfo, errs := accountService.GetAccount(context.Background(), c.config, c.accountsFetcher, request.Account)
+		if len(errs) > 0 {
+			return usersync.Request{}, privacy.Policies{}, combineErrors(errs)
+		}
+		if request.Limit == 0 {
+			request.Limit = accountInfo.CookieSync.DefaultLimit
+		}
+		if request.Limit == 0 || request.Limit > accountInfo.CookieSync.MaxLimit {
+			request.Limit = accountInfo.CookieSync.MaxLimit
+		}
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply endpoints/cookie_sync.go:145]: Also, can we move logic between lines 135 and 149 into its own function? It seems that the `parseRequest(r *http.Request)` function is already big enough. Let me know your thoughts.

**Review by guscarreon — COMMENTED** [review]
Looks great. Last couple of comments

**Comment by AlexBVolcy** [comment]
I put the main part of the implementation into its own function, `setLimit`, which gets called during `parseRequest`, to respond to Gus' comment! I've also added a new `TestSetLimit` function that covers all code from `setLimit`. Some of the limit related edge case tests have been moved from `TestCookieSyncParseRequest`to `TestSetLimit`, since it makes more sense for them to live there.

**Review by guscarreon — APPROVED** [review]
LGTM. Thanks for addressing the feedback.

**Review by VeronikaSolovei9 — APPROVED** [review]
